Update Smithsonian Unit code checker DAG to alert to Slack

stacimc opened the pull request:
## Fixes
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, please consider opening one before creating this pull request. -->
Fixes #420

## Description
<!-- Concisely describe what the pull request does. -->
<!-- Add screenshots, videos, or other media to show the problem and the solution when appropriate. -->
On main, we have a DAG called `check_new_smithsonian_unit_codes_workflow` which can only be run manually (it's not on a schedule), which compares the configuration in the `SMITHSONIAN_SUB_PROVIDERS` dictionary with the unit codes available in the Smithsonian API and raises an exception if they have fallen out of sync. It also creates and populates a DB table with the codes that need to be synced, which appears to be for legacy purposes and is no longer used anywhere.

This PR:

* Removes the two outdated unit codes from the Smithsonian sub provider configuration
* Entirely removes the `check_new_smithsonian_unit_codes_workflow` DAG and moves the logic into the Smithsonian provider script itself, so that it will fail fast if the unit code configuration needs to be changed.
* Removes logic for creating/updating DB table and instead reports the needed changes to Slack in a message like this:

<img width="644" alt="Screen Shot 2022-04-04 at 3 15 24 PM" src="https://user-images.githubusercontent.com/63313398/161640943-0eb8cf21-6d64-4c1c-a631-0fc01de5f66e.png">

## Notes
* Currently if the unit code validation fails, we'll get the Slack message with the error, but the `load_complete` Slack message will _also_ fire. There's an issue to address that: #465


## Testing Instructions
<!-- Give steps for the reviewer to verify that this PR fixes the problem; or delete this section entirely. -->
**Prereqs to test Slack:**
* You'll need to make sure you have the Slack connection set up for the Alerts channel. In your .env, that's `AIRFLOW_CONN_SLACK_ALERTS=<slack_connection_url>`. Troubleshooting tip: make sure it is formatted correctly with double-prepended 'https' and URL encoding (eg `"https://https%3A%2F%2Fhooks.slack.com%2Fservices%2everythingelse"`)
* You'll need to override the default behavior of omitting Slack messages locally. Go into the Airflow UI and select Admin > Variables, then add a variable named `slack_message_override` with value `true`.

Test:
* Check out this branch and run the Smithsonian provider script. With the changes in this PR, it should run successfully.
    * Watch the logs for the `pull_data` task and verify it starts writing records to disk, then mark the task as successful.
* Manually update the `SMITHSONIAN_SUB_PROVIDERS` dictionary in `provider_details.py` to make the unit codes out of sync. Try adding in non-existing codes ("FOO" etc), and deleting existing codes.
    * Run the provider script again and verify that the `pull_data` task fails and the error is sent to Slack. 

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. -->
- [ ] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [ ] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [ ] My commit messages follow [best practices][best_practices].
- [ ] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

stacimc commented:
Rebased to fix tests!

I keep going back and forth about whether to omit outdated unit codes or intentionally configure all of them. I don't think there's necessarily a _wrong_ choice so curious what others think.

Another consideration is the scheduling of the `check_new_smithsonian_unit_codes_workflow` DAG. Currently it's not scheduled at all and has to be run manually. The Smithsonian DAG itself is run weekly, and we'd like the checker to catch new codes beforehand. We _could_ run it daily (since it's such a small task).

On the other hand and especially now that we've removed the database aspect to it, I actually think it might be a good idea to eliminate the checker DAG entirely and simply perform the unit code validation as a first step in the Smithsonian's `main` function. That way we alert to Slack and fail early, wasting no time. Thoughts?

AetherUnbound commented on openverse_catalog/dags/common/loader/smithsonian_unit_codes.py:
```diff
@@ -61,74 +28,34 @@ def get_new_and_outdated_unit_codes(unit_code_set, sub_prov_dict=SUB_PROVIDERS):
-    initialise_unit_code_table(postgres_conn_id, unit_code_table)
-
-    postgres = PostgresHook(postgres_conn_id=postgres_conn_id)
+    if bool(new_unit_codes) or bool(outdated_unit_codes):
+        message = "\n*Updates needed to the SMITHSONIAN_SUB_PROVIDERS dictionary*:\n\n"
```
Since we're using `bool(<value>)` a few times, it might be worth doing assignment expressions here?

```suggestion
    if (has_new_codes := bool(new_unit_codes)) or (has_outdated_codes := bool(outdated_unit_codes)):
        message = "\n*Updates needed to the SMITHSONIAN_SUB_PROVIDERS dictionary*:\n\n"
```

And then use `has_<new|outdated>_codes` below

stacimc replied:
I generally agree, but in this case I think if `has_new_codes` is True, then the conditional will immediately be evaluated to True and `has_outdated_codes` won't get assigned 🤔 Since we're only evaluating `bool(<value>)` twice per value, I'm not sure it's worth working around that.

AetherUnbound replied:
WOAH, okay I definitely didn't consider that! Good to know for the future :memo:

AetherUnbound commented on openverse_catalog/dags/common/loader/smithsonian_unit_codes.py:
```diff
@@ -61,74 +28,34 @@ def get_new_and_outdated_unit_codes(unit_code_set, sub_prov_dict=SUB_PROVIDERS):
-        )
+        if bool(new_unit_codes):
+            codes_string = "\n".join(f"  - `{code}`" for code in new_unit_codes)
+            message += "New unit codes must be added:\n"
+            message += codes_string
+            message += "\n"
```
Might be worth doing a docstring here:
```suggestion
            message += f"""\
New unit codes must be added:
{codes_string}
"""
```

stacimc replied:
We could do that! I was annoyed by what the lack of whitespace does to the formatting, but obviously not a big deal 😄

AetherUnbound replied:
I hear it :sweat_smile: We could use [`textwrap.dedent`](https://docs.python.org/3/library/textwrap.html#textwrap.dedent) as well, but sometimes that can misbehave if you're formatting in a variable that includes newlines (because those lines don't have the same number of starting spaces).

AetherUnbound commented on openverse_catalog/dags/common/slack.py:
```diff
@@ -253,6 +253,21 @@ def should_send_message(http_conn_id=SLACK_NOTIFICATIONS_CONN_ID):
 
+def send_alert(
+    text: str,
+    username: str = "Airflow",
```
Maybe a bit more specific here? Not sure if that really benefits us here, but thought I'd throw it out there :woman_shrugging: 
```suggestion
    username: str = "Airflow Alert",
```

AetherUnbound reviewed (commented):
I love deleting code! :partying_face: The approach here looks good and I love the tests. Do you mind adding a screenshot of an example message if possible?

I think I would also be totally okay with moving this logic _into_ the provider script. It seems like it's a necessary step for its operation, so having it as part of the operational code makes sense to me. I suspect that many of these provider scripts were written in such a way that they could be tested on bare-metal, outside the docker stack. We've now (for better or worse) made running these within the docker stack a requirement, at least for providers that require an API key. I'm sure that the separation of these two steps was an attempt to make it easier to run the provider script locally - but our DevX is better now so I think it's okay that we run/test them within the stack :smile:

AetherUnbound commented on openverse_catalog/dags/providers/provider_api_scripts/smithsonian.py:
```diff
@@ -186,6 +190,48 @@ def _gather_unit_sample(unit, sample_dir, retries=RETRIES, endpoint=SEARCH_ENDPO
 
 
+def get_new_and_outdated_unit_codes(unit_code_set, sub_prov_dict=SUB_PROVIDERS):
+    sub_provider_unit_code_set = set()
+
+    for sub_prov, unit_code_sub_set in sub_prov_dict.items():
+        sub_provider_unit_code_set = sub_provider_unit_code_set.union(unit_code_sub_set)
```
If I'm interpreting this correctly, I think we can just do something like the following:
```python
[ins] In [1]: d = {1: {1, 2, 3}, 2: {3, 4, 5}, 3: {8, 9}}

[ins] In [2]: set().union(*d.values())
Out[2]: {1, 2, 3, 4, 5, 8, 9}
```

```suggestion
    all_sub_provider_unit_codes = set().union(*sub_prov_dict.values())
```

stacimc replied:
For sure -- this is just a copy/paste of the old logic 😄 I'll update the types while I'm at it!

AetherUnbound replied:
D'oh! Okay that makes a lot of sense :sweat_smile:

AetherUnbound commented on openverse_catalog/dags/providers/provider_api_scripts/smithsonian.py:
```diff
@@ -186,6 +190,48 @@ def _gather_unit_sample(unit, sample_dir, retries=RETRIES, endpoint=SEARCH_ENDPO
             break
 
 
+def get_new_and_outdated_unit_codes(unit_code_set, sub_prov_dict=SUB_PROVIDERS):
```
Nit: A number of these variables are suffixed with the type (e.g. `_set`, `_dict`). It might be better to remove the suffixes and add type annotations to the function instead.

AetherUnbound commented on tests/dags/providers/provider_api_scripts/test_smithsonian.py:
```diff
@@ -48,6 +49,47 @@ def test_get_hash_prefixes_with_len_one():
+    with patch.object(
+        si,
+        "get_new_and_outdated_unit_codes",
+        return_value=(set(new_unit_codes), set(outdated_unit_codes)),
```
Could `new_unit_codes` and `outdated_unit_codes` be defined as sets in the parametrization?

AetherUnbound reviewed (approved):
So much simpler!!! I have a few comments, but nothing to stop a merge :smile: :rocket:

stacimc commented:
Sorry about the linting errors, there was a mismatch between my local and CI. Rebased and it seems to be good now 👍

AetherUnbound commented on tests/dags/providers/provider_api_scripts/test_smithsonian.py:
```diff
@@ -73,11 +73,11 @@ def test_validate_unit_codes_from_api_raises_exception(
         with pytest.raises(
             AirflowException,
-            match="^\n\*Updates needed to the SMITHSONIAN_SUB_PROVIDERS dictionary\**",
+            match="^\n\\*Updates needed to the SMITHSONIAN_SUB_PROVIDERS dictionary\\**",
```
I think you could maybe use a [raw string](https://www.askpython.com/python/string/python-raw-strings) here so you don't have to escape the escape:
```suggestion
            match=r"^\n\*Updates needed to the SMITHSONIAN_SUB_PROVIDERS dictionary\**",
```

sarayourfriend commented:
@obulat and @dhruvkb can one of y'all review this PR soon?

obulat reviewed (approved):
I love the simplification of the Smithsonian data collection! 
Would it be possible to just add / remove the codes in the loader itself programmatically, without requiring the manual fixes?

PS. I meant to comment on the code, but approved instead. I need some help actually testing the Slack integration. Where do I get the connection URL?

AetherUnbound commented:
> PS. I meant to comment on the code, but approved instead. I need some help actually testing the Slack integration. Where do I get the connection URL?

It should be in our secret store! Note that there are some annoying things you have to do in order to get the connection set up correctly: 
https://github.com/WordPress/openverse-catalog/blob/c928589b96138efd823c4aca099cbc4baf020e62/env.template#L49-L54

We're working on improving those in #433 🙂

obulat reviewed (approved):
Thank you for all the detailed instructions for testing this PR. I got it to send a Slack message, as you describe.

stacimc commented:
Rebased for merge conflicts :)

> Would it be possible to just add / remove the codes in the loader itself programmatically, without requiring the manual fixes?

There actually used to be a DB table being populated by the checker (which I removed), but it was just used to store codes that need to be added/removed -- you still had to inspect the table and manually make changes to the constant in `provider_details.py`. We could reintroduce that and use it in the loader _instead_ of the configuration file, but I prefer the current setup. The loader will fail immediately, so there's no wasted time. I also like getting a chance to review the changes before automatically updating 😄 If the codes change more frequently than we expect and require a ton of manual intervention, we should revisit the process!
